src/tf_service/client.py

```diff
@@ -42,47 +108,99 @@ def __init__(self, server_node_name, keepalive_period=None):
-    def wait_for_server(self, timeout=rospy.Duration(-1)):
+    def wait_for_server(self, timeout: rospy.Duration) -> bool:
         """
         Block until the server is ready to respond to requests and reconnect.
 
         :param timeout: Time to wait for the server.
         :return: True if the server is ready, false otherwise.
         :rtype: bool
         """
-        return self.client.wait_for_server(timeout)
+        if self.is_connected():
+            return True
+        return self.reconnect(timeout)
 
-    @translate_exceptions
-    def is_connected(self):
-        return self.client.is_connected()
+    def is_connected(self) -> bool:
+        return _is_valid(self._can_transform_client) and _is_valid(
+            self._lookup_transform_client
+        )
 
-    @translate_exceptions
-    def reconnect(self, timeout=rospy.Duration(10)):
-        return self.client.reconnect(timeout)
+    def reconnect(self, timeout=rospy.Duration(10)) -> bool:
+        if self.is_connected():
+            return True
 
-    @translate_exceptions
-    def async_reconnect(self, timeout=rospy.Duration(10)):
-        return self.client.async_reconnect(timeout)
+        if self._reconnection_mutex.locked():
+            rospy.logwarn("Already reconnecting to tf_service server.")
+            return False
+
+        with self._reconnection_mutex:
+            try:
+                rospy.loginfo(
+                    "Waiting for %s to become available.",
+                    self._can_transform_client.resolved_name,
+                )
+                self._can_transform_client.wait_for_service(timeout)
+            except rospy.ROSException:
+                rospy.logerr("Failed to connect to tf_service server.")
+                return False
+
+            self._can_transform_client = rospy.ServiceProxy(
+                self._can_transform_client.resolved_name, CanTransform, persistent=True
+            )
+            _init_transport(self._can_transform_client)
+            self._lookup_transform_client = rospy.ServiceProxy(
+                self._lookup_transform_client.resolved_name,
+                LookupTransform,
+                persistent=True,
+            )
+            _init_transport(self._lookup_transform_client)
+        rospy.loginfo(
+            "Connected to services %s & %s",
+            self._can_transform_client.resolved_name,
+            self._lookup_transform_client.resolved_name,
+        )
+        return True
+
+    async def async_reconnect(self, timeout=None):
+        if self._reconnection_mutex.locked():
+            rospy.logdebug("Already asynchronously reconnecting to server.")
+            return
+        rospy.loginfo("Asynchronously trying to reconnect to tf_service server.")
+        self.reconnect(timeout)
 
     def _keepalive_callback(self, _):
         if not self.is_connected():
-            self.async_reconnect(rospy.Duration(-1))
+            asyncio.run(self.async_reconnect())
```

Comment:
This is probably not a clean use of asyncio. Works however. Check if needs to be changed, alternatively consider even removing this async feature completely from the Python client.

Reply:
Actually works fine for our use case. Would create a conflict if someone has this optional reconnect feature enabled here, plus another global asyncio.run().